### PR TITLE
Add rules for a better taint coverage

### PR DESCRIPTION
**Pull request by ErebusZ**

### CODE REVIEW

**Review by BlueSquare1 — CHANGES_REQUESTED** [review]
Please update unit tests

File: app/src/main/java/co/ostorlab/insecure_app/bugs/calls/SqlInjection.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    public String getDescription() {
+        return "The application sends unsanitized queries to the sql server";
+    }
+}
+
+
+class ExecuteQueryTask extends AsyncTask<String, Void, String> {
+
+    @Override
+    protected String doInBackground(String... params) {
```
Comment (PiranhaSa) [inline app/src/main/java/co/ostorlab/insecure_app/bugs/calls/SqlInjection.java:34]: is this supposed to be used somewhere ?

File: app/src/main/java/co/ostorlab/insecure_app/bugs/calls/XmlInjection.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        serializer.setOutput(fos, "UTF-8");
+
+        // Start the XML document
+        serializer.startDocument(null, Boolean.TRUE);
+
+        // Start the root element
+        serializer.startTag(null, "root");
+
+        // Inject the sanitized user input into the XML
+        serializer.text(user_input);
```
Comment (3asm) [inline app/src/main/java/co/ostorlab/insecure_app/bugs/calls/XmlInjection.java]: .text is not vulnerable. You need to append and parse.

File: app/src/main/java/co/ostorlab/insecure_app/bugs/calls/InsecureTemplateInjection.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        return "insecure template injection.";
+    }
+
+    @Override
+    public void run(String user_input) throws Exception {
+
+        TemplateRenderer templateRenderer = new TemplateRenderer();
+
+        String templateFileName = "template.mustache";
+        String renderedTemplate = templateRenderer.renderTemplate(appContext.getApplicationContext(), templateFileName, user_input);
```
Comment (3asm) [inline app/src/main/java/co/ostorlab/insecure_app/bugs/calls/InsecureTemplateInjection.java]: This is not vulnerable, you need to concatenate to trigger detection.

**Review by 3asm — CHANGES_REQUESTED** [review]
Please add a secure implementation to each to make sure the engine do not report them.

File: app/src/main/java/co/ostorlab/insecure_app/bugs/calls/InsecureTemplateInjection.java (hunk 3 above)
Comment (BlueSquare1) [reply app/src/main/java/co/ostorlab/insecure_app/bugs/calls/InsecureTemplateInjection.java]: Regardless `mustache` is logicless meaning it's not vulnerable to template injection by design.

**Review by BlueSquare1 — CHANGES_REQUESTED** [review]

**Comment by 3asm** [comment]
Please add test case of Wifi PII API.

File: app/src/main/java/co/ostorlab/insecure_app/bugs/calls/InsecureTemplateInjection.java [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    public String getDescription() {
+        return "Insecure template injection.";
+    }
+
+    @Override
+    public void run(String user_input) throws Exception {
+        TemplateRenderer templateRenderer = new TemplateRenderer();
+
+        String templateFileName = "template.ftl";
+        String renderedTemplate = templateRenderer.renderTemplate(appContext, templateFileName, user_input);
```
Comment (BlueSquare1) [inline app/src/main/java/co/ostorlab/insecure_app/bugs/calls/InsecureTemplateInjection.java]: Was this tested to make sure it actually triggers template injection?

Comment (3asm) [inline app/src/main/java/co/ostorlab/insecure_app/bugs/calls/InsecureTemplateInjection.java]: This is not vulnerable. The template should be concatenated from input.

File: app/src/main/java/co/ostorlab/insecure_app/bugs/calls/InsecureWifiApi.java [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+    }
+
+    @Override
+    public void run(String user_input) throws Exception {
+        WifiManager wifiManager = (WifiManager) appContext.getSystemService(Context.WIFI_SERVICE);
+
+        if (wifiManager != null && wifiManager.isWifiEnabled()){
+            wifiManager.getConnectionInfo();
+            wifiManager.getConfiguredNetworks();
+            wifiManager.getDhcpInfo();
```
Comment (3asm) [inline app/src/main/java/co/ostorlab/insecure_app/bugs/calls/InsecureWifiApi.java]: Should call `getBSSID` API.

File: app/src/main/java/co/ostorlab/insecure_app/bugs/calls/XmlInjection.java (hunk 2 above)
Comment (3asm) [reply app/src/main/java/co/ostorlab/insecure_app/bugs/calls/XmlInjection.java]: Ping.

**Review by 3asm — CHANGES_REQUESTED** [review]

File: app/src/main/java/co/ostorlab/insecure_app/bugs/calls/XmlInjection.java (hunk 2 above)
Comment (BlueSquare1) [reply app/src/main/java/co/ostorlab/insecure_app/bugs/calls/XmlInjection.java]: +1 you need to concatenate user input then parse it as xml document

**Review by BlueSquare1 — DISMISSED** [review]

File: app/src/main/java/co/ostorlab/insecure_app/bugs/calls/InsecureWifiApi.java (hunk 5 above)
Comment (ErebusZ) [reply app/src/main/java/co/ostorlab/insecure_app/bugs/calls/InsecureWifiApi.java]: the getConnectionInfo also returns the BSSID

File: app/src/main/java/co/ostorlab/insecure_app/bugs/calls/InsecureTemplateInjection.java [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+    }
+}
+
+class TemplateRenderer {
+
+    public String renderTemplate(Context context, String templateFileName, String user_input) {
```
Comment (3asm) [inline app/src/main/java/co/ostorlab/insecure_app/bugs/calls/InsecureTemplateInjection.java]: What is the point of `templateFileName`.

File: app/src/main/java/co/ostorlab/insecure_app/bugs/calls/InsecureTemplateInjection.java [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+            StringTemplateLoader templateLoader = new StringTemplateLoader();
+
+            String template_content = "<!DOCTYPE html><html><body><h1>Hello, " + user_input + "!</h1></body></html>";
+            templateLoader.putTemplate("hardcodedTemplate.ftl", template_content);
+            cfg.setTemplateLoader(templateLoader);
+            cfg.setDefaultEncoding("UTF-8");
+            cfg.setTemplateExceptionHandler(TemplateExceptionHandler.RETHROW_HANDLER);
+
+
+            // Get the template
```
Comment (3asm) [inline app/src/main/java/co/ostorlab/insecure_app/bugs/calls/InsecureTemplateInjection.java]: Oh really, that is a helpful comment. Please remove useless comments please.

File: app/src/main/java/co/ostorlab/insecure_app/bugs/calls/InsecureTemplateInjection.java [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+            Template template = cfg.getTemplate("hardcodedTemplate.ftl");
+
+            // Create the data model
+            StringWriter writer = new StringWriter();
+            template.process(null, writer);
+
+
```
Comment (3asm) [inline app/src/main/java/co/ostorlab/insecure_app/bugs/calls/InsecureTemplateInjection.java]: Fix line spacing.

File: app/src/main/java/co/ostorlab/insecure_app/bugs/calls/RegisterBroadcastReceiverDynamically.java [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+    @Override
+    public String getDescription() {
+        return "Register a broadcast receiver dynamically.";
+    }
+
+    @Override
+    public void run(String user_input) throws Exception{
+        BroadcastReceiver dynamicReceiver = new MyDynamicReceiver();
+        IntentFilter filter = new IntentFilter();
+        appContext.registerReceiver(dynamicReceiver, filter);
```
Comment (3asm) [inline app/src/main/java/co/ostorlab/insecure_app/bugs/calls/RegisterBroadcastReceiverDynamically.java]: Where is the flag that mark this as exported.

File: app/src/main/java/co/ostorlab/insecure_app/bugs/calls/RegisterBroadcastReceiverDynamically.java [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+    }
+    private class MyDynamicReceiver extends BroadcastReceiver {
+        @Override
+        public void onReceive(Context context, Intent intent) {
+            // Handle the received broadcast
```
Comment (3asm) [inline app/src/main/java/co/ostorlab/insecure_app/bugs/calls/RegisterBroadcastReceiverDynamically.java]: Logic?

File: app/src/main/java/co/ostorlab/insecure_app/bugs/calls/SqlInjection.java [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+public class SqlInjection extends BugRule {
+
+    @Override
+    public void run(String user_input) throws Exception {
+
+        if (!user_input.isEmpty()) {
+            new ExecuteQueryTask().execute(user_input);
+        }
+        else{
+            new ExecuteQueryTask().execute("INSERT INTO accounts(name, amount) VALUES(?, ?)");
```
Comment (3asm) [inline app/src/main/java/co/ostorlab/insecure_app/bugs/calls/SqlInjection.java]: Where is name and amount. This is not correct.

File: app/src/main/java/co/ostorlab/insecure_app/bugs/calls/SqlInjection.java [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+
+            // Execute the query
+            ResultSet resultSet = statement.executeQuery(query);
+
+            // Process the result set (replace this with your own logic)
+            while (resultSet.next()) {
+                result += resultSet.getString(1) + "\n";
+            }
+
+            // Close resources
```
Comment (3asm) [inline app/src/main/java/co/ostorlab/insecure_app/bugs/calls/SqlInjection.java]: Remove useless comments.

File: app/src/main/java/co/ostorlab/insecure_app/bugs/calls/XmlInjection.java [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+import org.xmlpull.v1.XmlSerializer;
+
+
+import co.ostorlab.insecure_app.BugRule;
+
+public class XmlInjection extends BugRule {
+
+    @Override
+    public void run(String user_input) throws Exception {
+
```
Comment (3asm) [inline app/src/main/java/co/ostorlab/insecure_app/bugs/calls/XmlInjection.java]: Fix code formatting.

File: app/src/main/res/raw/template.ftl [hunk 14]
```diff
@@ -0,0 +1,13 @@
+<!DOCTYPE html>
```
Comment (3asm) [inline app/src/main/res/raw/template.ftl]: This is no longer used.

**Review by 3asm — CHANGES_REQUESTED** [review]

File: app/src/main/java/co/ostorlab/insecure_app/bugs/calls/InsecureWifiApi.java (hunk 5 above)
Comment (BlueSquare1) [reply app/src/main/java/co/ostorlab/insecure_app/bugs/calls/InsecureWifiApi.java]: well in that case we should call `getBSSID` as well for better coverage.

Comment (3asm) [reply app/src/main/java/co/ostorlab/insecure_app/bugs/calls/InsecureWifiApi.java]: It returns a `WifiInfo` class.

File: app/src/main/java/co/ostorlab/insecure_app/bugs/calls/InsecureWifiApi.java [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+    public String getDescription() {
+        return "Insecure wifi api.";
+    }
+
+    @Override
+    public void run(String user_input) throws Exception {
+        WifiManager wifiManager = (WifiManager) appContext.getSystemService(Context.WIFI_SERVICE);
+
+        if (wifiManager != null && wifiManager.isWifiEnabled()){
+            wifiManager.getConnectionInfo();
```
Comment (3asm) [inline app/src/main/java/co/ostorlab/insecure_app/bugs/calls/InsecureWifiApi.java]:
```suggestion
            wifiManager.getConnectionInfo().getBSSID();
```

**Review by 3asm — APPROVED** [review]

**Pull request merged**